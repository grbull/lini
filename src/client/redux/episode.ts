/* eslint-disable no-param-reassign */
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { api } from '../utils/api';

type EpisodeState = {
  isLoading: boolean;
  error?: string;
  data?: EpisodeDto;
};

const initialState: EpisodeState = {
  isLoading: false,
};

export const get = createAsyncThunk<EpisodeDto, number>('episode/get', (id) =>
  api.episode.get(id)
);

export const episode = createSlice({
  name: 'episode',
  initialState,
  reducers: {
    clear: (state) => {
      state.data = undefined;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(get.pending, (state) => {
      state.isLoading = true;
    });
    builder.addCase(get.fulfilled, (state, { payload }) => {
      state.isLoading = false;
      state.data = payload;
    });
    builder.addCase(get.rejected, (state, { error }) => {
      state.isLoading = false;
      state.error = error.message;
    });
  },
});

export const episodeActions = { ...episode.actions, get };
