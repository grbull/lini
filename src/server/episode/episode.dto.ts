import { ShowDto } from '../show/show.dto';

export class EpisodeDto {
  id!: string;
  name!: string;
  season!: number;
  number!: number;
  type!: string;
  airdate?: string;
  airtime?: string;
  airstamp?: string;
  runtime?: number;
  imageMedium?: string;
  imageOriginal?: string;
  summary?: string;
  show!: ShowDto;
}

export class ScheduleDto {
  future!: EpisodeDto[];
  past!: EpisodeDto[];
}
