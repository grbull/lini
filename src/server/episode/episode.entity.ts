import { Column, Entity, ManyToOne } from 'typeorm';

import { ShowEntity } from '../show/show.entity';

@Entity('episode')
export class EpisodeEntity {
  @Column('bigint', { primary: true })
  id!: string;

  @Column('text')
  name!: string;

  @Column('integer')
  season!: number;

  @Column('integer')
  number!: number;

  @Column('text')
  type!: string;

  @Column('date', { nullable: true })
  airdate!: string | null;

  @Column('time', { nullable: true })
  airtime!: string | null;

  @Column('timestamptz', { nullable: true })
  airstamp!: string | null;

  @Column('integer', { nullable: true })
  runtime!: number | null;

  @Column('text', { nullable: true })
  imageMedium!: string | null;

  @Column('text', { nullable: true })
  imageOriginal!: string | null;

  @Column('text', { nullable: true })
  summary!: string | null;

  @ManyToOne(() => ShowEntity, (show) => show.episodes)
  show!: number;
}
