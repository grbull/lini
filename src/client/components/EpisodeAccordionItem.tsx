import cn from 'classnames';
import React, { ReactElement } from 'react';
import { Link } from 'react-router-dom';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { dateToLocaleDate } from '../utils/date';
import { formatEpisodeNumber } from '../utils/formatEpisodeNumber';

interface Props {
  episodes: EpisodeDto[];
  isExpanded: boolean;
  season: number;
  setExpanded: (row: number) => void;
}

export function EpisodeAccordionItem({
  episodes,
  isExpanded,
  season,
  setExpanded,
}: Props): ReactElement {
  function clickHandler(): void {
    setExpanded(isExpanded ? 0 : season);
    // TODO: Scroll to bottom
  }

  return (
    // eslint-disable-next-line jsx-a11y/click-events-have-key-events
    <div
      className={cn('py-1')}
      key={season}
      onClick={clickHandler}
      role="button"
      tabIndex={season}
    >
      <div
        className={cn('flex', 'justify-between', 'items-center', {
          'font-bold': isExpanded,
        })}
      >
        <span>Season {season}</span>
        <span>{episodes.length} Episodes</span>
      </div>
      {isExpanded &&
        episodes.map((episode) => (
          <Link
            className={cn('py-1', 'flex', 'justify-between', 'items-center')}
            key={episode.number}
            to={`/episode/${episode.id}`}
          >
            <span>
              <span className={cn('mr-1', 'text-gray-400')}>
                E{formatEpisodeNumber(episode.number)}
              </span>
              <span>{episode.name}</span>
            </span>
            <span>
              {episode.airstamp ? dateToLocaleDate(episode.airstamp) : 'N/A'}
            </span>
          </Link>
        ))}
    </div>
  );
}
