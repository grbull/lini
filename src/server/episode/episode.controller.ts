import {
  Controller,
  Get,
  Param,
  UseGuards,
  UseInterceptors,
} from '@nestjs/common';

import { User } from '../common/decorators/user.decorator';
import { AuthGuard } from '../common/guards/auth.guard';
import { LoggingInterceptor } from '../common/interceptors/logging.interceptor';
import { UserEntity } from '../user/user.entity';
import { EpisodeEntity } from './episode.entity';
import { EpisodeService } from './episode.service';

@Controller('episode')
@UseGuards(AuthGuard)
@UseInterceptors(LoggingInterceptor)
export class EpisodeController {
  constructor(private readonly episodeService: EpisodeService) {}

  @Get('schedule')
  public async schedule(
    @User() user: UserEntity
  ): Promise<{ future: EpisodeEntity[]; past: EpisodeEntity[] }> {
    const future = await this.episodeService.getFuture(user);
    const past = await this.episodeService.getPast(user);
    return { future, past };
  }

  @Get(':id')
  public getOne(@Param('id') id: number): Promise<EpisodeEntity> {
    return this.episodeService.getOne(id);
  }
}
