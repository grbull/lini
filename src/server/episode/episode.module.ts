import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { LoggerModule } from '../logger/logger.module';
import { SubscriptionModule } from '../subscription/subscription.module';
import { EpisodeController } from './episode.controller';
import { EpisodeEntity } from './episode.entity';
import { EpisodeService } from './episode.service';

@Module({
  imports: [
    LoggerModule,
    TypeOrmModule.forFeature([EpisodeEntity]),
    SubscriptionModule,
  ],
  providers: [EpisodeService],
  controllers: [EpisodeController],
  exports: [EpisodeService],
})
export class EpisodeModule {}
