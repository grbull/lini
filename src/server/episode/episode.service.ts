import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Between, Repository } from 'typeorm';

import { LoggerService } from '../logger/logger.service';
import { SubscriptionService } from '../subscription/subscription.service';
import { TvMazeEpisode } from '../tvmaze/tvmaze.types';
import { tvMazeEpisodeToEntity } from '../tvmaze/tvmaze.utils';
import { UserEntity } from '../user/user.entity';
import { EpisodeEntity } from './episode.entity';

@Injectable()
export class EpisodeService {
  constructor(
    private readonly loggerService: LoggerService,
    @InjectRepository(EpisodeEntity)
    private readonly episodeRepository: Repository<EpisodeEntity>,
    private readonly subscriptionService: SubscriptionService
  ) {
    this.loggerService.setContext('EpisodeService');
  }

  public async getOne(id: number): Promise<EpisodeEntity> {
    try {
      return await this.episodeRepository
        .createQueryBuilder('episode')
        .leftJoinAndSelect('episode.show', 'show')
        .where('episode.id = :id', { id })
        .getOneOrFail();
    } catch {
      throw new NotFoundException();
    }
  }

  // public getByShow(id: number): Promise<EpisodeEntity[]> {
  //   return this.episodeRepository
  //     .createQueryBuilder('episode')
  //     .where('episode.show.id = :id', { id })
  //     .orderBy('episode.airdate', 'ASC')
  //     .getMany();
  // }

  // public async getUpcoming(user: UserEntity): Promise<EpisodeEntity[]> {
  //   const subscriptions = await this.subscriptionService.getAll(user);
  //   let upcomingEpisodes: EpisodeEntity[] = [];

  //   const oneMonthAhead = new Date();
  //   oneMonthAhead.setMonth(new Date().getMonth() + 1);

  //   for (const key of subscriptions) {
  //     const episodes = await this.episodeRepository.find({
  //       where: {
  //         // between now and 1 month ahead
  //         // airstamp: Raw(
  //         //   (alias) => `${alias} < '${oneMonthAhead.toISOString()}'`
  //         // ),
  //         airstamp: Between(new Date(), oneMonthAhead),
  //         show: key.show,
  //       },
  //     });
  //     console.log(episodes);
  //     if (episodes) {
  //       upcomingEpisodes = [...upcomingEpisodes, ...episodes];
  //     }
  //   }
  //   return upcomingEpisodes;
  // }

  public async getFuture(user: UserEntity): Promise<EpisodeEntity[]> {
    const subscriptions = await this.subscriptionService.getAll(user);

    const oneMonthAhead = new Date();
    oneMonthAhead.setMonth(new Date().getMonth() + 1);

    const findOptions = subscriptions.map((subscription) => {
      return {
        airstamp: Between(new Date(), oneMonthAhead),
        show: subscription.show,
      };
    });

    return this.episodeRepository.find({
      where: findOptions,
      order: {
        airstamp: 'ASC',
        number: 'ASC',
      },
      relations: ['show'],
    });
  }

  public async getPast(user: UserEntity): Promise<EpisodeEntity[]> {
    const subscriptions = await this.subscriptionService.getAll(user);

    const oneMonthBehind = new Date();
    oneMonthBehind.setMonth(new Date().getMonth() - 1);

    const findOptions = subscriptions.map((subscription) => {
      return {
        airstamp: Between(oneMonthBehind, new Date()),
        show: subscription.show,
      };
    });

    return this.episodeRepository.find({
      where: findOptions,
      order: {
        airstamp: 'DESC',
        number: 'DESC',
      },
      relations: ['show'],
    });
  }

  // The belowe is used for seeding and updating
  public createOrUpdateMany(
    episodes: TvMazeEpisode[],
    showId: number
  ): Promise<EpisodeEntity[]> {
    const episodeEntities = episodes.map((episode) =>
      tvMazeEpisodeToEntity(episode, showId)
    );

    return this.episodeRepository.save(episodeEntities, { chunk: 1000 });
  }

  // used by notification service
  public findByTime(
    showId: number,
    startTime: Date,
    endTime: Date
  ): Promise<EpisodeEntity[]> {
    return this.episodeRepository.find({
      where: { show: showId, airstamp: Between(startTime, endTime) },
    });
  }
}
