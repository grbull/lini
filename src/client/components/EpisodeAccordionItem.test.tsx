/**
 * @jest-environment jsdom
 */
/* eslint-disable init-declarations */
import '@testing-library/jest-dom/extend-expect';

import { fireEvent, render } from '@testing-library/react';
import { createMemoryHistory, MemoryHistory } from 'history';
import React from 'react';
import { Router } from 'react-router-dom';

import { EpisodeDto } from '../../server/episode/episode.dto';
import { EpisodeAccordionItem } from './EpisodeAccordionItem';

const episodes: Omit<EpisodeDto, 'show'>[] = [
  {
    id: '1627015',
    name: 'Red Moon',
    season: 1,
    number: 1,
    type: 'regular',
    airdate: '2019-11-01',
    airtime: null,
    airstamp: '2019-11-01T12:00:00.000Z',
    runtime: 65,
    imageMedium: '/uploads/images/medium_landscape/218/545490.jpg',
    imageOriginal: '/uploads/images/original_untouched/218/545490.jpg',
    summary:
      'NASA is in crisis as the Soviets land the first man on the moon in 1969, the beginning of an alternate history.',
  },
];

// Unfortunately I can't get 100% line coverage because of optional chaining.

describe('EpisodeAccordionItem Component', () => {
  let history: MemoryHistory<unknown>;
  let handleExpanded: (n: number) => void;

  beforeEach(() => {
    handleExpanded = jest.fn();
    history = createMemoryHistory();
    window.HTMLElement.prototype.scrollIntoView = jest.fn();
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(
      <EpisodeAccordionItem
        episodes={episodes}
        isExpanded={false}
        season={1}
        setExpanded={handleExpanded}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('handles being clicked when expanded', () => {
    const { getByRole } = render(
      <Router history={history}>
        <EpisodeAccordionItem
          episodes={episodes}
          isExpanded
          season={1}
          setExpanded={handleExpanded}
        />
      </Router>
    );
    fireEvent.click(getByRole('button'));
    expect(handleExpanded).toHaveBeenCalled();
    expect(getByRole('button').scrollIntoView).toBeCalledTimes(0);
  });

  it('handles being clicked when not expanded', () => {
    const { getByRole } = render(
      <EpisodeAccordionItem
        episodes={episodes}
        isExpanded={false}
        season={1}
        setExpanded={handleExpanded}
      />
    );
    fireEvent.click(getByRole('button'));
    expect(handleExpanded).toBeCalledTimes(1);
    expect(getByRole('button').scrollIntoView).toBeCalledTimes(1);
  });

  it('episodes should have a link to their page', () => {
    const { getByText } = render(
      <Router history={history}>
        <EpisodeAccordionItem
          episodes={episodes}
          isExpanded
          season={1}
          setExpanded={handleExpanded}
        />
      </Router>
    );
    expect(getByText('Red Moon').closest('a')).toHaveAttribute(
      'href',
      '/episode/1627015'
    );
  });

  it('should render episodes without an airstamp', () => {
    const { getByText } = render(
      <Router history={history}>
        <EpisodeAccordionItem
          episodes={[{ ...episodes[0], airstamp: undefined }]}
          isExpanded
          season={1}
          setExpanded={handleExpanded}
        />
      </Router>
    );
    expect(getByText('N/A')).toBeTruthy();
  });
});
