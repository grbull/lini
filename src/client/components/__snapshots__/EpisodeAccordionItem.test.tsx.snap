// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpisodeAccordionItem Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="py-1"
    role="button"
    tabindex="1"
  >
    <div
      class="flex justify-between items-center"
    >
      <span>
        Season 1
      </span>
      <span>
        1 Episodes
      </span>
    </div>
  </div>
</DocumentFragment>
`;
